Assets/SEE/GraphProviders/VCSGraphProvider.cs

```diff
@@ -0,0 +1,421 @@
+        /// <returns>the graph.</returns>
+        static Graph GetVCSGraph(Dictionary<string, bool> pathGlobbing, string repositoryPath, string commitID)
+        {
+            string[] pathSegments = repositoryPath.Split(Path.DirectorySeparatorChar);
+
+            Graph graph = new(repositoryPath, pathSegments[^1]);
+
+            // The main directory.
+            GraphUtils.NewNode(graph, pathSegments[^1], "directory", pathSegments[^1]);
+
+            IEnumerable<string> includedFiles = pathGlobbing
+                .Where(path => path.Value == true)
+                .Select(path => path.Key);
+
+            IEnumerable<string> excludedFiles = pathGlobbing
+                .Where(path => path.Value == false)
+                .Select(path => path.Key);
+
+            using (Repository repo = new(repositoryPath))
+            {
+                LibGit2Sharp.Tree tree = repo.Lookup<Commit>(commitID).Tree;
+                // Get all files using "git ls-tree -r <CommitID> --name-only".
+                IEnumerable<string> files;
+                if (includedFiles.Any() && !string.IsNullOrEmpty(includedFiles.First()))
+                {
+                    files = ListTree(tree).Where(path => includedFiles.Contains(Path.GetExtension(path)));
+                }
+                else if (excludedFiles.Any())
+                {
+                    files = ListTree(tree).Where(path => !excludedFiles.Contains(Path.GetExtension(path)));
+                }
+                else
+                {
+                    files = ListTree(tree);
+                }
+
+                // Build the graph structure.
+                foreach (string filePath in files.Where(path => !string.IsNullOrEmpty(path)))
+                {
+                    string[] filePathSegments = filePath.Split(Path.AltDirectorySeparatorChar);
+                    // Files in the main directory.
+                    if (filePathSegments.Length == 1)
+                    {
+                        graph.GetNode(pathSegments[^1]).AddChild(GraphUtils.NewNode(graph, filePath, "file", filePath));
+                    }
+                    // Other directorys/files.
+                    else
+                    {
+                        GraphUtils.BuildGraphFromPath(filePath, null, null, graph, graph.GetNode(pathSegments[^1]));
+                    }
+                }
+                AddMetricsToNode(graph, repo, commitID);
+            }
+
+            return graph;
+        }
+
+        /// <summary>
+        /// Gets the paths from a repository at the time of a given commitID.
+        /// It is equivalent to "git ls-tree --name-only -r commitID"
+        /// </summary>
+        /// <param name="tree">The tree of the given commit.</param>
+        /// <returns>a list of paths.</returns>
+        static IEnumerable<string> ListTree(LibGit2Sharp.Tree tree)
+        {
+            List<string> fileList = new();
+
+            foreach (TreeEntry entry in tree)
```

Comment:
If you tested this on bigger repositories (like SEE)—is this where the main bottleneck lies or is it somewhere else? If performance was never a problem even for a repository of SEE's size, you can ignore this comment, but otherwise, it could be a good idea to make this `async` in order not to block the UI (if you need assistance for this, I can help). In that case, you could also use `changePercentage` to display a progress bar.

Reply:
The building of the graph itself is not the problem with bigger repositories, but the drawing is. I actually dont know how the drawing works. But maybe you have an idea. Also the method you pointed at, just gets all the paths from the repository with git ls-tree --name-only -r commitID, that is a realtively fast operation. But the parsing and building of the graph is where most of the work happens. (GraphUtils.BuildGraphFromPath()).

Reply:
I've implemented the asynchronous graph drawing mentioned here in #741. As for the other suggestion (making the loop through file paths asynchronous), it might make sense to do a quick performance comparison between the current method and making it run on the thread pool for bigger graphs.

Reply:
It took me 4 milliseconds to load SEE (46338 nodes), I dont know if that would be overengineering.

Reply:
In that case, we can definitely ignore it. Switching to the thread pool has its own overhead which would dwarf the time required to load SEE, so we shouldn't do it IMO.